cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/file/FileSetDataset.java

```diff
@@ -251,4 +253,27 @@ public String getOutputFormatClassName() {
   private String getFileSystemPath(Location loc) {
     return loc.toURI().getPath();
   }
+
+  @Override
+  public void onSuccess() throws DataSetException {
+    // nothing needed to do on success
+  }
+
+  @Override
+  public void onFailure() throws DataSetException {
+    Location outputLocation = getOutputLocation();
+    // If there is no output path, it is either using DynamicPartitioner or the job would have failed.
+    // Either way, we can't do much here.
+    if (outputLocation == null) {
+      return;
+    }
+
+    try {
+      if (!outputLocation.delete(true)) {
```

Comment:
Now I am not sure what it means if delete() returns false. Seems that it may also indicate an error. Perhaps check if it exists.